[max] Fix for concurrency issue and small cleanups

gedejong opened the pull request:
- Fix deadlocks caused by synchronized (without timeout).
- Various small fixes and cleanups

gedejong commented:
Hi @kaikreuzer, could you have a look at this PR? Thank in advance.

kaikreuzer commented:
@marcelrv Would be the right person to review.

kaikreuzer commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/discovery/MaxCubeBridgeDiscovery.java:
```diff
@@ -44,12 +44,12 @@
     private final static int SEARCH_TIME = 15;
 
-    private final Logger logger = LoggerFactory.getLogger(MaxCubeBridgeDiscovery.class);
+    private static final Logger LOGGER = LoggerFactory.getLogger(MaxCubeBridgeDiscovery.class);
```
please revert this, see [C.3](http://docs.openhab.org/developers/development/guidelines.html#c-language-levels-and-libraries).

gedejong commented:
@kaikreuzer thanks for the quick review. I wasn't aware of the static vs. non-static issue on logging and believed non-static loggers were generally a mistake (not anymore). I reverted to non-static loggers and cleaned up some other parts where loggers weren't used or didn't have the right modifiers.

I'm looking at the original  `org.openhab.binding.max.internal.handler.MaxCubeBridgeHandler.sendCubeCommand(CubeCommand)` now. I believe I fixed a bug with the `maxRequestsPerConnection` if-condition. If the maximum request is reached, it will not go to the `else` block, but still return a successful true, thus popping the command off the stack without processing.

Also, I was wondering: wouldn't it be easier to have a separate thread be signalled when a command is available to be send? Now it seems to be polling for commands on the queue, perhaps because of throttling?

gedejong commented:
@kaikreuzer please wait a bit with this PR. Yesterday I had another dead-lock with the new code. Probably the scheduler thread-pool is full (perhaps caused by another binding). I hacked a separate thread-pool in the max-binding to see if that's the reason. Running this code now for a while and no problems yet.

marcelrv commented:
Hallo @gedejong 
Thank you for your contribution. With the original code, did you experience many dead-locks?
If the pool is full, is the solution not to increase the pool size for your install, or do you believe it is specifically tied to the MAX! binding? If not specifically tied to the max biding, we should not create new pools see design guideline D.2 http://www.eclipse.org/smarthome/documentation/development/guidelines.html 

Wrt to your question, indeed I introduced throttling the commands to the cube to eliminate structural issues with the command sending in the earlier versions of the binding. There are 2 main reasons. 1) The cube sometimes becomes unstable if too many commands are send in short period (reboots/hangs). 2) Some users set the temperature with a slider, where the slider may generate many commands for the in-between values. The throttling & queue replacement logic ensures we are not sending all these commands to the cube. Otherwise the duty cycle would be full in very short time.

Can you push the version you believe should be right, I'' review it

marcelrv commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/command/UdpCubeCommand.java:
```diff
@@ -152,15 +152,6 @@ private void receiveUdpCommandResponse() {
         } catch (IOException e) {
             logger.debug("IO error during MAX! Cube response: {}", e.getMessage());
             commandRunning = false;
-        } finally {
```
Is there a reason why the port should not be closed and socket is cleared ?

gedejong replied:
I changed it to the Java 7 
``` try (closable = SomethingCloseable) { block; } ``` syntax. This takes care of checking for nullability as well as exception handling.

marcelrv replied:
thx. cool, did not yet know that syntax.

marcelrv commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/discovery/MaxCubeBridgeDiscovery.java:
```diff
@@ -165,15 +152,6 @@ private void receiveDiscoveryMessage() {
         } catch (IOException e) {
             logger.debug("IO error during MAX! Cube discovery: {}", e.getMessage());
             discoveryRunning = false;
-        } finally {
```
Same here, why should the port not be closed after usage

marcelrv commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -92,33 +96,33 @@ public MaxCubeBridgeHandler(Bridge br) {
+    private final Logger logger = LoggerFactory.getLogger(MaxCubeBridgeHandler.class);
 
     /** The refresh interval which is used to poll given MAX! Cube */
-    private long refreshInterval = 30;
```
If you move the variable, please move the comment along

gedejong replied:
Oops, thanks, will fix that.

marcelrv commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -181,9 +189,19 @@ public void handleCommand(ChannelUID channelUID, Command command) {
     @Override
     public void dispose() {
         logger.debug("Handler disposed.");
-        stopAutomaticRefresh();
```
In what situation do you see the scheduler start & stop fail?

gedejong replied:
Generally after running for around 12 hours.

gedejong replied:
This was caused by another binding. Since then I've had this code (except for the last commit, which removes some of the excess ReentrantLocks) running without any problems.

marcelrv commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -147,21 +151,25 @@ public MaxCubeBridgeHandler(Bridge br) {
 
     private Set<DeviceStatusListener> deviceStatusListeners = new CopyOnWriteArraySet<>();
 
+    private final Lock pollingJobLock = new ReentrantLock();
```
Are you sure you are seeing locks due to the synchronized?

gedejong replied:
It was probably caused be another process eating up the threads in the scheduler. At first, I thought the scheduler was local to each binding (which seems more sensible to me).

I analysed by starting with the synchronised blocks. Since the class is already rather large and stateful, I decided to protect the shared variables separately. This gives fidelity on time-outs. These two variables were previously protected using a synchronized keyword. It would technically be possible for two external threads to start these jobs, both checking for nullability on the scheduled future and starting two separate jobs.

marcelrv replied:
Starting the handler twice would be an significant issue of the framework. Handlers are only started once per thing.
We should avoid to create complexity to work around (unlikely) possible bugs of the framework.

MAX Binding should be able to run for months stable (and can...), so if in your case it runs only for 12h indeed we need to find the root cause and fix that.
There must be a combination of factors driving the instability you experience, let's find & fix that

The main purpose of the synchronized block is to avoid we are sending a command and send another one prior receiving the response of the first. 
Only in the more exceptional cases (e.g. during new device inclusion)  there is a significant delay between the sending of the command and the response. In all other cases, response is immediately, and the commands are consumed from the queue with a fixed delay, basically excluding the chance for clashes.

gedejong replied:
I'll revert the locking code here.

gedejong commented:
> With the original code, did you experience many dead-locks?
If the pool is full, is the solution not to increase the pool size for your install, or do you believe it is specifically tied to the MAX! binding? If not specifically tied to the max biding, we should not create new pools see design guideline D.2 http://www.eclipse.org/smarthome/documentation/development/guidelines.html

Yes, it happened regularly. Perhaps my approach was wrong. I thought I could look at the code and fix it, without debugging. Currently I have a VisualVM connected with the instance. This should give me some extra information.

Nevertheless, the changes made in the PR seem to have a positive effect on stability and speed. The threads that are started are almost constantly parked, without holding any locks. 

I'll debug the 2.1.0-SNAPSHOT implementation and compare it with the PR and see what could be causing the deadlocks after all. 

Also, on a completely other note: I'm currently porting the DSMR binding to OH 2.0 (so it has discovery and Things for the various meters). I'm a bit afraid of double-work. Is there a certain procedure?

marcelrv commented:
@gedejong 
I'm happy with all of the cleanup you did in all of the files. Thanks for that.
I also think the lock as an alternative to the synchronized on the sendcommand is fine.

I'm not happy with the own threads mechanism. Unless we can truly pinpoint that this is the rootcause of your issues. 

I think the dutycycle monitor is a good idea. I'm not happy with the current implementation of it.
I think that if the lock is there or the duty cycle is too high we should avoid pulling more commands from the queue and trying to send it. I think in the current implementation after 20 seconds the sendcommand is lost. (esp the inclusion is dangerous as it locks for 60 seconds and it does not go via the queue)
Likewise it should be avoided that due to the dutycycle monitor no command is send, as there are commands that don't 'consume' the duty cycle (all commands that are for the cube, hence not air commands)

gedejong commented:
Hi Marcel, thanks for your comments. I monitored with visualvm and the root-cause was a blocking Runnable from the network-binding (the default time-out is 5 seconds, and it blocks...) I'm feeling a bit bad for not having used VisualVM before on this... Sorry for this.

So, I just reverted the custom scheduling pool.

About the duty cycle. When the duty-cycle is high, I suggest putting a timeout on the `Condition` and either closing the socket (so the refresh task will reconnect with the cube, thus retrieving the new duty-cycle, thus possibly unblocking the condition), or sending a command which will report the duty cycle as well (I don't know which one). I tried the first option, and it seems to work nicely. 

I also changed the timeout from 20 seconds to 120 seconds, which should be long enough for the device inclusion.

marcelrv commented:
hallo @gedejong were you planning to make another update to address the implementation issues/concerns? 
Otherwise suggest we push all of the changes outside of the cubehandler through

gedejong commented:
Ok, I simplified the PR a bit. I have had the previous version running for the last months, without any problems. Just tested the new version with some scenarios, such as the duty-cycle check, which works well. I don't really know why Jenkins is failing though...

martinvw commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -450,67 +555,59 @@ public void clearDeviceList() {
+            socketClose(); // reconnect on next execution
+            return false;
+        } catch (IOException e) {
+            logger.warn("IO error occurred during execution: {}", e);
```
e does not need a placeholder

martinvw commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -520,41 +617,34 @@ private synchronized boolean sendCubeCommand(CubeCommand command) {
-        boolean cont = true;
-        while (cont) {
+    private void readLines(String terminator) throws IOException {
+        while (true) {
```
That could take a while, while !interupted maybe?

gedejong replied:
There are various return statements in the block. The previous logic was, subjectively, harder to read.

martinvw replied:
Yes, but I was actually refering to [`Thread.isInterrupted()`](https://docs.oracle.com/javase/7/docs/api/java/lang/Thread.html#isInterrupted()) if some other process interupts your thread. But looking close at the code its just a while input is received.

martinvw commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -520,41 +617,34 @@ private synchronized boolean sendCubeCommand(CubeCommand command) {
-            } else {
-                cont = false;
+            } catch (Exception e) {
+                logger.info("Error while handling message block: '{}' from MAX! Cube lan gateway: {}:", raw, ipAddress,
```
Non matching number of placeholders

martinvw commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -520,41 +617,34 @@ private synchronized boolean sendCubeCommand(CubeCommand command) {
                 }
-            } else {
-                cont = false;
+            } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

gedejong replied:
I agree. I couldn't clean this because I didn't know which exception was actually the possible culprit.

gedejong replied:
(it's a refactor, didn't write the original code)

martinvw commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -566,110 +656,126 @@ private void readliness(String terminator) throws IOException {
+                        if (di != null) {
+                            deviceStatusListener.onDeviceConfigUpdate(getThing(), di);
+                        }
+                    } catch (NullPointerException e) {
```
Please not

gedejong replied:
wasn't my code. I was afraid to refactor it, because I didn't know what exactly could NPE...

martinvw commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -566,110 +656,126 @@ private void readliness(String terminator) throws IOException {
+        } else if (message.getType() == MessageType.F) {
+            setProperties((F_Message) message);
+        }
+    }
```
This method became very big 100+ lines of code...

gedejong replied:
Yeah, I had already written an Enum with a lambda and then I remembered we don't support Java 8 yet... Completely agreed, it'll be split.

martinvw replied:
We do have Java 8, since some weeks 👍

gedejong replied:
Yay! Now we have to refactor everything again :P

martinvw replied:
Next time I will say: :-P

> This method became very big 100+ lines of code... Did you know we have Java 8 since a few weeks.

gedejong commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -129,7 +130,11 @@ public MaxCubeBridgeHandler(Bridge br) {
     private boolean propertiesSet = false;
```
Personally, I'd prefer to have each variable explicitly initialised, either in the constructor or as an argument.

martinvw replied:
I'm pretty sure they don't, the only one is maybe about `@NonNull` marked fields (I'm interested in a reference for that). It's valid practice to leave them out, only local variables require initilaization. Note also that a lot of fields are already not initialised via the constructor. 

While checking the code again, I saw something which troubles me more :-)

The constructor is above the fields, can you move it below the fields.

martinvw replied:
See also https://blogs.oracle.com/tor/entry/code_advice_14_don_t

gedejong replied:
Well, I have to say I'm in firm disagreement with Oracle here, but if it is code-convention, that'll take priority.

gedejong commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -181,8 +180,13 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+        try {
+            stopAutomaticRefresh();
+        } catch (InterruptedException e) {
+            logger.error("Could not stop automatic refresh", e);
```
Thanks, I was unaware of the intricacies. I've become used to an actor and stream based concurrency control (Akka/Scala).

gedejong commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -306,60 +312,159 @@ private synchronized void startAutomaticRefresh() {
-        if (sendCommandJob == null || sendCommandJob.isCancelled()) {
-            sendCommandJob = scheduler.scheduleWithFixedDelay(sendCommandRunnable, 0, sendCommandInterval,
-                    TimeUnit.SECONDS);
+        if (queueConsumerThread == null || !queueConsumerThread.isAlive()) {
```
The scheduler used a polling method, waking up a Thread (cpu) and caused a noticeable delay. The `queueConsumerThread` will wait() on reading the queue.

martinvw replied:
Up to @kaikreuzer

kaikreuzer replied:
ok for me

gedejong commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -306,60 +312,159 @@ private synchronized void startAutomaticRefresh() {
+                    Thread.sleep(50);
+                }
+            } catch (InterruptedException e) {
+                logger.info("Stopping queueConsumer");
```
This is probably not an issue here, since it's the last statement of the thread.

gedejong commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -306,60 +312,159 @@ private synchronized void startAutomaticRefresh() {
+                            socket.close();
+                        }
+                    } catch (IOException e) {
+                        logger.warn("Could not close socket", e);
```
agreed, the stack should have been printed by the logger statement above.

gedejong commented on addons/binding/org.openhab.binding.max/META-INF/MANIFEST.MF:
```diff
@@ -22,5 +22,7 @@ Import-Package: com.google.common.base;version="10.0.1",
  org.slf4j
 Service-Component: OSGI-INF/*
 Export-Package: org.openhab.binding.max
+Require-Bundle: org.eclipse.smarthome.core,
```
Ok, OSGi is still new for me.

marcelrv commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -933,4 +999,8 @@ private void updateCubeState() {
         updateState(new ChannelUID(getThing().getUID(), CHANNEL_DUTY_CYCLE), new DecimalType(dutyCycle));
     }
+
+    private boolean hasExcessDutyCycle() {
```
This probably needs to become a public one.
The MaxDevicesHandler should not initiate the actuals refreshing if this condition exists.

marcelrv commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -566,110 +656,156 @@ private void readliness(String terminator) throws IOException {
-                    if (c != null) {
-                        configurations.remove(c);
-                    }
+    private void processNMessage(N_Message nMessage) {
```
Better to merge processNMessage &  newInclusionDeviceFound.
For consistency call it processNMessage

gedejong commented:
Hi @marcelrv, thanks for the thorough review. I decided to make the code a bit more consistent and checked all source files for unnecessary initialization, incorrect generics and such. Also ran findbugs to see if I missed anything more. Some internationalisation issues, naming issues, incorrect scoping, unnecessary javadoc and smaller things have been changed as well.

martinvw commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/message/L_Message.java:
```diff
@@ -18,6 +18,8 @@
 import org.openhab.binding.max.internal.device.DeviceConfiguration;
 import org.slf4j.Logger;
 
+import com.google.common.base.Charsets;
```
Since Java 7 there is StandardCharsets please take a look at that, I prefer JDK defaults over non-standard stuff.

martinvw commented:
@gedejong sorry for reviewing in multiple steps :-)

As we are slowly going in the direction of a new release it would be nice to include these changes as well. So if you have time it would be great if we could wrap up this PR. If you need any help or have questions just let us know.

And of course thanks for all the work you already did.

marcelrv commented:
Also, I was wondering: wouldn't it be easier to have a separate thread be signalled when a command is available to be send? Now it seems to be polling for commands on the queue, perhaps because of throttling?

This is indeed for throttling. The cubes tend to hang or worse, there's are reports of corruption. 
I suspect the data rate of sending commands (+firmware bug) to be the issue.

martinvw commented:
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.0.0:compile (default-compile) on project org.openhab.binding.max.test: Compilation failure: Compilation failure:
[ERROR] /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java:[49,54] 1. ERROR in /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java (at line 49)
[ERROR] ArrayList<DeviceInformation> allDevicesInformation = message.devices;
[ERROR] ^^^^^^^^^^^^^^^
[ERROR] Type mismatch: cannot convert from List<DeviceInformation> to ArrayList<DeviceInformation>
[ERROR] 
[ERROR] /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java:[65,54] 2. ERROR in /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java (at line 65)
[ERROR] ArrayList<DeviceInformation> allDevicesInformation = message.devices;
[ERROR] ^^^^^^^^^^^^^^^
[ERROR] Type mismatch: cannot convert from List<DeviceInformation> to ArrayList<DeviceInformation>
[ERROR] 
[ERROR] /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java:[73,54] 3. ERROR in /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java (at line 73)
[ERROR] ArrayList<DeviceInformation> allDevicesInformation = message.devices;
[ERROR] ^^^^^^^^^^^^^^^
[ERROR] Type mismatch: cannot convert from List<DeviceInformation> to ArrayList<DeviceInformation>
[ERROR] 
[ERROR] /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java:[81,54] 4. ERROR in /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java (at line 81)
[ERROR] ArrayList<DeviceInformation> allDevicesInformation = message.devices;
[ERROR] ^^^^^^^^^^^^^^^
[ERROR] Type mismatch: cannot convert from List<DeviceInformation> to ArrayList<DeviceInformation>
[ERROR] 
[ERROR] /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java:[89,54] 5. ERROR in /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java (at line 89)
[ERROR] ArrayList<DeviceInformation> allDevicesInformation = message.devices;
[ERROR] ^^^^^^^^^^^^^^^
[ERROR] Type mismatch: cannot convert from List<DeviceInformation> to ArrayList<DeviceInformation>
[ERROR] 
[ERROR] /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java:[97,46] 6. ERROR in /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.max.test/src/test/java/org/openhab/binding/max/internal/message/M_MessageTest.java (at line 97)
[ERROR] ArrayList<RoomInformation> roomInformation = message.rooms;
[ERROR] ^^^^^^^^^^^^^
[ERROR] Type mismatch: cannot convert from List<RoomInformation> to ArrayList<RoomInformation>
[ERROR] 
[ERROR] Found 6 errors and 0 warnings.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To se

martinvw commented:
@gedejong thanks for all your efforts already, do you plan to finish these changes?

gedejong commented:
And I'm sorry I didn't pick it up earlier. I had a slightly busier than usual month in the professional sphere and had to drop the evening coding for a couple of weeks. However, I can pick it up again and I'll do that this week!

gedejong commented:
@martinvw well, I worked on the Charsets and your other CR comments. I've had the code running for a while on my own Openhab 2 installation, and it seems to work fine. The moment in the year doesn't really allow for much testing though, heh.

martinvw commented:
@kaikreuzer can you do a quick final review? Thanks!

kaikreuzer commented on addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/handler/MaxCubeBridgeHandler.java:
```diff
@@ -450,67 +554,59 @@ public void clearDeviceList() {
+
+        } catch (ConnectException e) {
+            logger.warn("Connection timed out on {} port {}", ipAddress, port);
+            socketClose(); // reconnect on next execution
```
Looks like it - according to the guidelines, all logging here should only be debug and the ThingStatus should be set instead.

kaikreuzer commented:
Sorry, I just noticed that I cannot merge yet - the pullapprove hook correctly complains about: "Signed-off-by" needed on 14 commits

@gedejong Could you please squash your commits and add a Signed-off-by footer to the resulting commit? See https://github.com/openhab/openhab2-addons/blob/master/CONTRIBUTING.md#sign-your-work for details!

kaikreuzer commented:
@gedejong Are you still around?

martinvw commented:
@gedejong any updates from you?

wborn commented:
It would be nice if we could finally merge this PR since #3506 is waiting on it. 
Could you resolve the merge conflict and squash/sign off the commits @gedejong as requested?
If that's not possible somehow, please drop us a comment and tell us how we should proceed with this.

wborn commented:
Since @gedejong seems to be MIA for almost 1 year. What would be the best way to continue with Max Binding PRs @kaikreuzer and @martinvw?

* Squash and fix the sign offs in this PR ourselves (possibly using a forked branch). Would that violate the DCO?
* Continue merging other Max PRs such as #3506. Though that will only increase the merge conflicts of this PR.

marcelrv commented:
who can advice here? @kaikreuzer ?

May I suggest we indeed to Continue merging other Max PRs such as #3506. Though that will only increase the merge conflicts of this PR. as suggested with @wborn 

Assuming @gedejong does not squash as he is missing in action can we take the contributions and have signoff by someone else? I would not mind reworking it

kaikreuzer commented:
As I know that dutch are fine people, I would be ok to assume that @gedejong worked on this PR under the prerequisite of the DCO and he simply missed to add the sign-off.
So let us merge this and get it out of the way. We meanwhile also have one file in conflict, so squashing won't be enough, we also need a rebase.

Shall I take care of it?

martinvw commented:
That would be great!

kaikreuzer commented:
Done! If anything isn't working, please check my [conflict resolution commit](https://github.com/openhab/openhab2-addons/pull/1789/commits/1ccf3aa50e92363ff1b5c330adf51ff125623255) whether I did anything wrong there...

wborn commented:
Many thanks!

cweitkamp commented:
After rebasing #3506 I get some errors when I want to test it. First I was thinking it could be a problem of wrong resolved merge conflicts. Later I tried the latest master but it did not work either. I got some log entries like this:

2018-07-26 23:08:39.318 [WARN ] [nternal.handler.MaxCubeBridgeHandler] - IO error occurred during execution: Read timed out
2018-07-26 23:08:50.541 [WARN ] [nternal.handler.MaxCubeBridgeHandler] - IO error occurred during execution: Read timed out
2018-07-26 23:08:50.580 [WARN ] [nternal.handler.MaxCubeBridgeHandler] - Error sending command org.openhab.binding.max.internal.handler.SendCommand@1591ca8[id=0,channelUID=<null>,command=<null>,cubeCommand=org.openhab.binding.max.internal.command.F_Command@187835,serialNumber=Cube(XXXXXXXX),key=Cube(XXXXXXXX)-F_Command,commandText=Request NTP info] to MAX! Cube at IP: 192.168.1.20
2018-07-26 23:08:51.169 [WARN ] [nternal.handler.MaxCubeBridgeHandler] - IO error occurred during execution: Read timed out
2018-07-26 23:09:02.330 [WARN ] [nternal.handler.MaxCubeBridgeHandler] - IO error occurred during execution: Read timed out

I will do some more research with DEBUG mode later.

I found a topic in the community where others have reported the Sam issue after upgrading to latest snapshot: 
https://community.openhab.org/t/maxcubebridgehandler-io-error-occurred-during-execution-read-timed-out/48414

kaikreuzer commented:
Reverting is no solution, analysing and fixing the issue is.
As far as I understood @cweitkamp's comment, he planned to debug it and knowing his skills I am quite confident that he will find the issue and a solution.
